Bug: Stdlib functions return unsupported matrix types

## Summary

Several stdlib functions return nested arrays (matrices) even though matrix support is not officially implemented (see #47). This creates an inconsistent state where matrices work at runtime but cannot be declared with explicit types.

## Affected Functions

| Function | Returns |
|----------|---------|
| `arrays.zip()` | `[[T]]` - nested arrays |
| `arrays.chunk()` | `[[T]]` - nested arrays |
| `arrays.flatten()` | Consumes `[[T]]` |
| `bytes.split()` | `[[byte]]` - nested byte arrays |

## Current Behavior

```ez
// This WORKS - runtime creates and handles nested arrays
temp zipped = arrays.zip({1, 2, 3}, {4, 5, 6})
println(zipped)           // {{1, 4}, {2, 5}, {3, 6}}
println(zipped[0])        // {1, 4}
println(zipped[1][1])     // 5

// This FAILS - parser doesn't support [[type]] syntax
temp matrix [[int]] = {{1, 2}, {3, 4}}
// error[E2024]: expected type name, got [
```

## The Problem

1. **Runtime**: Fully supports nested arrays (creation, access, iteration, double-indexing)
2. **Stdlib**: Has functions that create/consume matrices
3. **Parser**: Cannot parse `[[type]]` declarations
4. **Type System**: No way to explicitly declare matrix types

This is inconsistent - users can create matrices via stdlib but can't declare them properly.

## Resolution

Add support for matrix parsing (`[[type]]` syntax) in the parser and type system. This will also resolve #47.

## Test File

See `tests/matrix_unsupported_test.ez` for reproduction.

## Related

- #47 (Matrix/2D array support)

--- a/pkg/parser/parser.go
+++ b/pkg/parser/parser.go
@@ -1655,26 +1655,33 @@ func (p *Parser) parseTypeName() string {
 		typeName := "["
 		p.nextToken() // move past [
 
-		// The element type should be an identifier
-		if !p.currentTokenMatches(IDENT) {
+		// The element type can be an identifier OR another array type (for matrices)
+		var elementType string
+		if p.currentTokenMatches(LBRACKET) {
+			// Nested array type [[type]] - recursive call
+			elementType = p.parseTypeName()
+			if elementType == "" {
+				return ""
+			}
+		} else if p.currentTokenMatches(IDENT) {
+			elementType = p.currentToken.Literal
+			// Check for qualified type name inside array: [module.TypeName]
+			if p.peekTokenMatches(DOT) {
+				p.nextToken() // consume the DOT
+				p.nextToken() // get the type name
+				if !p.currentTokenMatches(IDENT) {
+					msg := fmt.Sprintf("expected type name after '.', got %s", p.currentToken.Type)
+					p.errors = append(p.errors, msg)
+					p.addEZError(errors.E2024, msg, p.currentToken)
+					return ""
+				}
+				elementType = elementType + "." + p.currentToken.Literal
+			}
+		} else {
 			msg := fmt.Sprintf("expected type name, got %s", p.currentToken.Type)
 			p.errors = append(p.errors, msg)
 			p.addEZError(errors.E2024, msg, p.currentToken)
 			return ""
-		}
-		elementType := p.currentToken.Literal
-
-		// Check for qualified type name inside array: [module.TypeName]
-		if p.peekTokenMatches(DOT) {
-			p.nextToken() // consume the DOT
-			p.nextToken() // get the type name
-			if !p.currentTokenMatches(IDENT) {
-				msg := fmt.Sprintf("expected type name after '.', got %s", p.currentToken.Type)
-				p.errors = append(p.errors, msg)
-				p.addEZError(errors.E2024, msg, p.currentToken)
-				return ""
-			}
-			elementType = elementType + "." + p.currentToken.Literal
 		}
 		typeName += elementType
 

--- a/pkg/parser/parser_test.go
+++ b/pkg/parser/parser_test.go
@@ -1397,3 +1397,85 @@ func TestIsReservedName(t *testing.T) {
 		})
 	}
 }
+
+// ============================================================================
+// Matrix Type Parsing Tests
+// ============================================================================
+
+func TestMatrixTypeParsing(t *testing.T) {
+	tests := []struct {
+		name         string
+		input        string
+		expectedName string
+		typeName     string
+	}{
+		{"2D int array", "temp matrix [[int]] = {{1, 2}, {3, 4}}", "matrix", "[[int]]"},
+		{"2D string array", "temp strs [[string]] = {{\"a\", \"b\"}, {\"c\", \"d\"}}", "strs", "[[string]]"},
+		{"2D float array", "temp floats [[float]] = {{1.1, 2.2}, {3.3, 4.4}}", "floats", "[[float]]"},
+		{"2D bool array", "temp flags [[bool]] = {{true, false}, {false, true}}", "flags", "[[bool]]"},
+		{"3D int array", "temp cube [[[int]]] = {{{1, 2}, {3, 4}}, {{5, 6}, {7, 8}}}", "cube", "[[[int]]]"},
+		{"empty 2D array", "temp empty [[int]] = {}", "empty", "[[int]]"},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			program := parseProgram(t, tt.input)
+			if len(program.Statements) == 0 {
+				t.Fatalf("no statements parsed")
+			}
+			stmt, ok := program.Statements[0].(*VariableDeclaration)
+			if !ok {
+				t.Fatalf("not VariableDeclaration, got %T", program.Statements[0])
+			}
+			if stmt.Name.Value != tt.expectedName {
+				t.Errorf("expected name %q, got %q", tt.expectedName, stmt.Name.Value)
+			}
+			if stmt.TypeName != tt.typeName {
+				t.Errorf("expected typeName %q, got %q", tt.typeName, stmt.TypeName)
+			}
+		})
+	}
+}
+
+func TestMatrixInFunctionParameter(t *testing.T) {
+	input := `do processMatrix(m [[int]]) -> int {
+		return 0
+	}`
+	program := parseProgram(t, input)
+	fn := program.Statements[0].(*FunctionDeclaration)
+
+	if len(fn.Parameters) != 1 {
+		t.Fatalf("expected 1 parameter, got %d", len(fn.Parameters))
+	}
+	if fn.Parameters[0].TypeName != "[[int]]" {
+		t.Errorf("expected parameter type '[[int]]', got %q", fn.Parameters[0].TypeName)
+	}
+}
+
+func TestMatrixInFunctionReturn(t *testing.T) {
+	input := `do createMatrix() -> [[int]] {
+		return {{1, 2}, {3, 4}}
+	}`
+	program := parseProgram(t, input)
+	fn := program.Statements[0].(*FunctionDeclaration)
+
+	if len(fn.ReturnTypes) != 1 || fn.ReturnTypes[0] != "[[int]]" {
+		t.Errorf("expected return type '[[int]]', got %v", fn.ReturnTypes)
+	}
+}
+
+func TestMatrixInStructField(t *testing.T) {
+	input := `const Grid struct {
+		data [[int]]
+		rows int
+	}`
+	program := parseProgram(t, input)
+	stmt := program.Statements[0].(*StructDeclaration)
+
+	if len(stmt.Fields) != 2 {
+		t.Fatalf("expected 2 fields, got %d", len(stmt.Fields))
+	}
+	if stmt.Fields[0].TypeName != "[[int]]" {
+		t.Errorf("expected field type '[[int]]', got %q", stmt.Fields[0].TypeName)
+	}
+}

--- a/tests/comprehensive.ez
+++ b/tests/comprehensive.ez
@@ -112,6 +112,10 @@ do main() {
     total_passed += result.passed
     total_failed += result.failed
 
+    result = test_matrices()
+    total_passed += result.passed
+    total_failed += result.failed
+
     result = test_maps()
     total_passed += result.passed
     total_failed += result.failed
@@ -443,11 +447,75 @@ do test_arrays() -> TestResult {
     println("  after mutable[1] = 25: ${mutable}")
     passed += 1
 
-    // NOTE: 2D arrays [[int]] not supported - see GitHub issue #47
-
     // Fixed size array
     const fixed [int, 3] = {100, 200, 300}
     println("  fixed size array: ${fixed}")
+    passed += 1
+
+    println("  PASSED: ${passed}, FAILED: ${failed}")
+    return TestResult{passed: passed, failed: failed}
+}
+
+// ==================================================
+// MATRICES (2D Arrays)
+// ==================================================
+
+do test_matrices() -> TestResult {
+    print_section("Matrices (2D Arrays)")
+    temp passed int = 0
+    temp failed int = 0
+
+    // 2D int array declaration
+    temp int_matrix [[int]] = {{1, 2, 3}, {4, 5, 6}}
+    println("  [[int]] matrix: ${int_matrix}")
+    println("  int_matrix[0]: ${int_matrix[0]}")
+    println("  int_matrix[1][2]: ${int_matrix[1][2]}")
+    passed += 1
+
+    // 2D string array
+    temp str_matrix [[string]] = {{"a", "b"}, {"c", "d"}}
+    println("  [[string]] matrix: ${str_matrix}")
+    println("  str_matrix[1][0]: ${str_matrix[1][0]}")
+    passed += 1
+
+    // Empty 2D array
+    temp empty_matrix [[int]] = {}
+    println("  empty matrix length: ${len(empty_matrix)}")
+    passed += 1
+
+    // Matrix from arrays.zip
+    temp zipped [[int]] = arrays.zip({1, 2}, {3, 4})
+    println("  arrays.zip result: ${zipped}")
+    println("  zipped[0][1]: ${zipped[0][1]}")
+    passed += 1
+
+    // Matrix from arrays.chunk
+    temp chunked [[int]] = arrays.chunk({1, 2, 3, 4, 5, 6}, 2)
+    println("  arrays.chunk result: ${chunked}")
+    println("  chunked[1][0]: ${chunked[1][0]}")
+    passed += 1
+
+    // Flatten matrix back to 1D
+    temp nested [[int]] = {{10, 20}, {30, 40}}
+    temp flattened [int] = arrays.flatten(nested)
+    println("  arrays.flatten({{10,20},{30,40}}): ${flattened}")
+    passed += 1
+
+    // Iterate over matrix rows
+    temp rows [[int]] = {{1, 2}, {3, 4}, {5, 6}}
+    temp row_sum int = 0
+    for_each row in rows {
+        for_each val in row {
+            row_sum += val
+        }
+    }
+    println("  sum of all elements in matrix: ${row_sum}")
+    passed += 1
+
+    // 3D array (nested matrices)
+    temp cube [[[int]]] = {{{1, 2}, {3, 4}}, {{5, 6}, {7, 8}}}
+    println("  [[[int]]] 3D array: ${cube}")
+    println("  cube[1][0][1]: ${cube[1][0][1]}")
     passed += 1
 
     println("  PASSED: ${passed}, FAILED: ${failed}")

